File: server/src/routes/audio.ts
// /api/audio/*
//
// Serves the AudioAdapter shape for the web build. Files are cached under
// ALETHEIA_AUDIO_CACHE (default /tmp/aletheia-audio) — ephemeral on Railway
// but rebuilt on demand from upstream public-domain sources.
//
// Slug/filename validation mirrors src-tauri/src/audio.rs intentionally: the
// values land inside filesystem paths, so anything looser would be a path-
// traversal foot-gun. Keep these regexes in sync if the Rust side ever
// loosens its rules.

import { Router } from "express";
import type { Request, Response } from "express";
import { createReadStream, promises as fsp } from "fs";
import { Readable } from "stream";
import path from "path";

const ALLOWED_TRANSLATIONS = new Set(["en_bsb", "en_kjv", "en_web"]);

// Lowercase ascii + digits + underscore, 1..32 chars. Same as
// validate_slug() in audio.rs.
const SLUG_RE = /^[a-z0-9_]{1,32}$/;

// Filenames come from upstream URLs (LibriVox uses hyphens, parens, mixed
// case). Same as validate_filename() in audio.rs: alphanumeric + . _ - ( ),
// no path separators, no "..", must end with .mp3, 1..128 chars.
const FILENAME_RE = /^[A-Za-z0-9._\-()]{1,128}\.mp3$/;

interface ValidationOk {
  ok: true;
  translation: string;
  book: string;
  file: string;
}
interface ValidationErr {
  ok: false;
  status: number;
  error: string;
}
type Validation = ValidationOk | ValidationErr;

function validate(
  translation: unknown,
  book: unknown,
  file: unknown,
): Validation {
  if (typeof translation !== "string" || !ALLOWED_TRANSLATIONS.has(translation)) {
    return { ok: false, status: 400, error: `unsupported translation: ${translation}` };
  }
  if (typeof book !== "string" || !SLUG_RE.test(book)) {
    return { ok: false, status: 400, error: `invalid book slug: ${book}` };
  }
  if (file !== undefined) {
    if (typeof file !== "string" || file.includes("/") || file.includes("\\") || file.includes("..")) {
      return { ok: false, status: 400, error: `invalid filename: ${file}` };
    }
    if (!FILENAME_RE.test(file)) {
      return { ok: false, status: 400, error: `invalid filename: ${file}` };
    }
  }
  return { ok: true, translation, book, file: (file ?? "") as string };
}

function bookDir(cacheRoot: string, translation: string, book: string): string {
  return path.join(cacheRoot, translation, book);
}

function sourcePath(
  cacheRoot: string,
  translation: string,
  book: string,
  file: string,
): string {
  return path.join(bookDir(cacheRoot, translation, book), file);
}

function streamUrl(translation: string, book: string, file: string): string {
  return `/api/audio/stream/${translation}/${book}/${encodeURIComponent(file)}`;
}

async function fileExistsNonEmpty(p: string): Promise<boolean> {
  try {
    const st = await fsp.stat(p);
    return st.isFile() && st.size > 0;
  } catch {
    return false;
  }
}

export function audioRouter(cacheRoot: string): Router {
  const router = Router();

  router.get("/source-path", async (req, res) => {
    const v = validate(req.query.translation, req.query.book, req.query.file);
    if (!v.ok) {
      res.status(v.status).json({ error: v.error });
      return;
    }
    const p = sourcePath(cacheRoot, v.translation, v.book, v.file);
    const exists = await fileExistsNonEmpty(p);
    res.json({ url: streamUrl(v.translation, v.book, v.file), exists });
  });

  router.get("/book-sources", async (req, res) => {
    const v = validate(req.query.translation, req.query.book, undefined);
    if (!v.ok) {
      res.status(v.status).json({ error: v.error });
      return;
    }
    const dir = bookDir(cacheRoot, v.translation, v.book);
    let entries: string[] = [];
    try {
      const names = await fsp.readdir(dir);
      // Match the Rust side: only count .mp3 files that are present and
      // non-empty; skip half-written .part files and any zero-byte stragglers.
      const checked = await Promise.all(
        names
          .filter((n) => n.endsWith(".mp3"))
          .map(async (n) => ({ n, ok: await fileExistsNonEmpty(path.join(dir, n)) })),
      );
      entries = checked.filter((c) => c.ok).map((c) => c.n).sort();
    } catch (err: unknown) {
      if (!isENOENT(err)) {
        res.status(500).json({ error: `read_dir ${dir}: ${(err as Error).message}` });
        return;
      }
    }
    res.json(entries);
  });

  router.post("/download", async (req, res) => {
    const { translation, book, url, filename } = (req.body ?? {}) as {
      translation?: unknown;
      book?: unknown;
      url?: unknown;
      filename?: unknown;
    };
    const v = validate(translation, book, filename);
    if (!v.ok) {
      res.status(v.status).json({ error: v.error });
      return;
    }
    if (typeof url !== "string" || !/^https?:\/\//i.test(url)) {
      res.status(400).json({ error: `refusing non-http(s) URL: ${url}` });
      return;
    }
    const dest = sourcePath(cacheRoot, v.translation, v.book, v.file);
    const dir = path.dirname(dest);
    await fsp.mkdir(dir, { recursive: true });

    // Short-circuit if the file already exists. The Tauri code re-downloads
    // unconditionally; here we don't, because Railway's ephemeral disk means
    // the cache *is* the source of truth between deploys — a "download" call
    // after the file is already cached is almost certainly a frontend that
    // forgot to check source-path first.
    if (await fileExistsNonEmpty(dest)) {
      res.json({ url: streamUrl(v.translation, v.book, v.file) });
      return;
    }

    const part = `${dest}.part`;
    try {
      const upstream = await fetch(url, {
        headers: { "user-agent": "Aletheia/0.1 (https://github.com/Emessjay/aletheia)" },
      });
      if (!upstream.ok || !upstream.body) {
        res.status(502).json({ error: `GET ${url} returned HTTP ${upstream.status}` });
        return;
      }
      const out = (await import("fs")).createWriteStream(part);
      // The Web ReadableStream returned by undici needs adapting to Node's
      // stream API; Readable.fromWeb is the supported bridge.
      const nodeStream = Readable.fromWeb(upstream.body as unknown as Parameters<typeof Readable.fromWeb>[0]);
      await new Promise<void>((resolve, reject) => {
        nodeStream.pipe(out);
        nodeStream.on("error", reject);
        out.on("error", reject);
        out.on("finish", () => resolve());
      });
      await fsp.rename(part, dest);
      res.json({ url: streamUrl(v.translation, v.book, v.file) });
    } catch (err: unknown) {
      // Best-effort cleanup; ignore failure (the partial may already be gone).
      await fsp.unlink(part).catch(() => undefined);
      res.status(502).json({ error: `download failed: ${(err as Error).message}` });
    }
  });

  router.get("/stream/:translation/:book/:file", async (req, res) => {
    const v = validate(req.params.translation, req.params.book, req.params.file);
    if (!v.ok) {
      res.status(v.status).json({ error: v.error });
      return;
    }
    const p = sourcePath(cacheRoot, v.translation, v.book, v.file);
    let stat;
    try {
      stat = await fsp.stat(p);
    } catch {
      res.status(404).json({ error: "not found" });
      return;
    }
    if (!stat.isFile() || stat.size === 0) {
      res.status(404).json({ error: "not found" });
      return;
    }
    streamRange(req, res, p, stat.size);
  });

  return router;
}

function isENOENT(err: unknown): boolean {
  return typeof err === "object" && err !== null && (err as { code?: string }).code === "ENOENT";
}

// HTTP Range support. The browser <audio> tag fires a Range request whenever
// the user seeks, so without this every seek triggers a full re-download.
function streamRange(req: Request, res: Response, filePath: string, size: number): void {
  const range = req.headers.range;
  res.setHeader("accept-ranges", "bytes");
  res.setHeader("content-type", "audio/mpeg");

  if (!range) {
    res.setHeader("content-length", String(size));
    createReadStream(filePath).pipe(res);
    return;
  }

  const match = /^bytes=(\d*)-(\d*)$/.exec(range.trim());
  if (!match) {
    res.status(416).setHeader("content-range", `bytes */${size}`);
    res.end();
    return;
  }
  const startRaw = match[1];
  const endRaw = match[2];

  let start: number;
  let end: number;
  if (startRaw === "" && endRaw !== "") {
    // suffix-byte-range: last N bytes
    const suffix = Number.parseInt(endRaw, 10);
    if (!Number.isFinite(suffix) || suffix <= 0) {
      res.status(416).setHeader("content-range", `bytes */${size}`).end();
      return;
    }
    start = Math.max(0, size - suffix);
    end = size - 1;
  } else {
    start = startRaw === "" ? 0 : Number.parseInt(startRaw, 10);
    end = endRaw === "" ? size - 1 : Number.parseInt(endRaw, 10);
  }

  if (!Number.isFinite(start) || !Number.isFinite(end) || start > end || start < 0 || end >= size) {
    res.status(416).setHeader("content-range", `bytes */${size}`);
    res.end();
    return;
  }

  res.status(206);
  res.setHeader("content-range", `bytes ${start}-${end}/${size}`);
  res.setHeader("content-length", String(end - start + 1));
  createReadStream(filePath, { start, end }).pipe(res);
}

File: server-py/app/routes/audio.py
"""/api/audio/* — mirrors server/src/routes/audio.ts.

Slug/filename validation mirrors src-tauri/src/audio.rs intentionally: the
values land inside filesystem paths, so anything looser would be a path-
traversal foot-gun. Keep these regexes in sync if the Rust side ever loosens
its rules.
"""

from __future__ import annotations

import re
from pathlib import Path
from typing import Any
from urllib.parse import quote

import httpx
from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse, Response, StreamingResponse

from ..config import resolve_audio_cache

ALLOWED_TRANSLATIONS = {"en_bsb", "en_kjv", "en_web"}

SLUG_RE = re.compile(r"^[a-z0-9_]{1,32}$")
FILENAME_RE = re.compile(r"^[A-Za-z0-9._\-()]{1,128}\.mp3$")

# Match the existing Node value so upstream logs see the same identity.
USER_AGENT = "Aletheia/0.1 (https://github.com/Emessjay/aletheia)"

# Stream MP3s in 64 KiB chunks. Small enough to stay responsive on seek,
# large enough to keep syscall overhead off the hot path.
CHUNK_SIZE = 64 * 1024


def _err(message: str, status: int = 400) -> JSONResponse:
    return JSONResponse({"error": message}, status_code=status)


def _validate(translation: Any, book: Any, file: Any | None) -> tuple[str, str, str] | JSONResponse:
    if not isinstance(translation, str) or translation not in ALLOWED_TRANSLATIONS:
        return _err(f"unsupported translation: {translation}")
    if not isinstance(book, str) or not SLUG_RE.match(book):
        return _err(f"invalid book slug: {book}")
    if file is None:
        return translation, book, ""
    if not isinstance(file, str) or "/" in file or "\\" in file or ".." in file:
        return _err(f"invalid filename: {file}")
    if not FILENAME_RE.match(file):
        return _err(f"invalid filename: {file}")
    return translation, book, file


def _book_dir(cache_root: Path, translation: str, book: str) -> Path:
    return cache_root / translation / book


def _source_path(cache_root: Path, translation: str, book: str, file: str) -> Path:
    return _book_dir(cache_root, translation, book) / file


def _stream_url(translation: str, book: str, file: str) -> str:
    return f"/api/audio/stream/{translation}/{book}/{quote(file, safe='')}"


def _file_exists_nonempty(p: Path) -> bool:
    try:
        st = p.stat()
    except OSError:
        return False
    return st.st_size > 0 and p.is_file()


def audio_router() -> APIRouter:
    router = APIRouter()

    @router.get("/source-path")
    async def source_path(req: Request) -> Response:
        translation = req.query_params.get("translation")
        book = req.query_params.get("book")
        file = req.query_params.get("file")
        v = _validate(translation, book, file)
        if isinstance(v, JSONResponse):
            return v
        tr, bk, fl = v
        p = _source_path(resolve_audio_cache(), tr, bk, fl)
        return JSONResponse({"url": _stream_url(tr, bk, fl), "exists": _file_exists_nonempty(p)})

    @router.get("/book-sources")
    async def book_sources(req: Request) -> Response:
        translation = req.query_params.get("translation")
        book = req.query_params.get("book")
        v = _validate(translation, book, None)
        if isinstance(v, JSONResponse):
            return v
        tr, bk, _ = v
        dir_ = _book_dir(resolve_audio_cache(), tr, bk)
        if not dir_.exists():
            return JSONResponse([])
        try:
            names = sorted(
                p.name
                for p in dir_.iterdir()
                if p.name.endswith(".mp3") and _file_exists_nonempty(p)
            )
        except OSError as err:
            return _err(f"read_dir {dir_}: {err}", 500)
        return JSONResponse(names)

    @router.post("/download")
    async def download(req: Request) -> Response:
        try:
            body = await req.json()
        except Exception:
            return _err("invalid JSON body")
        if not isinstance(body, dict):
            return _err("body must be a JSON object")
        v = _validate(body.get("translation"), body.get("book"), body.get("filename"))
        if isinstance(v, JSONResponse):
            return v
        tr, bk, fl = v
        url = body.get("url")
        if not isinstance(url, str) or not re.match(r"^https?://", url, re.IGNORECASE):
            return _err(f"refusing non-http(s) URL: {url}")

        cache_root = resolve_audio_cache()
        dest = _source_path(cache_root, tr, bk, fl)
        dest.parent.mkdir(parents=True, exist_ok=True)

        # Short-circuit if already cached. The Tauri code re-downloads
        # unconditionally; here we don't, because Railway's ephemeral disk
        # means the cache *is* the source of truth between deploys.
        if _file_exists_nonempty(dest):
            return JSONResponse({"url": _stream_url(tr, bk, fl)})

        part = dest.with_suffix(dest.suffix + ".part")
        try:
            async with httpx.AsyncClient(
                follow_redirects=True,
                headers={"user-agent": USER_AGENT},
                timeout=httpx.Timeout(30.0, read=120.0),
            ) as client:
                async with client.stream("GET", url) as resp:
                    if resp.status_code >= 400:
                        return _err(f"GET {url} returned HTTP {resp.status_code}", 502)
                    with part.open("wb") as out:
                        async for chunk in resp.aiter_bytes(CHUNK_SIZE):
                            out.write(chunk)
            part.replace(dest)
            return JSONResponse({"url": _stream_url(tr, bk, fl)})
        except Exception as err:  # noqa: BLE001
            try:
                part.unlink()
            except OSError:
                pass
            return _err(f"download failed: {err}", 502)

    @router.get("/stream/{translation}/{book}/{file}")
    async def stream(translation: str, book: str, file: str, request: Request) -> Response:
        v = _validate(translation, book, file)
        if isinstance(v, JSONResponse):
            return v
        tr, bk, fl = v
        p = _source_path(resolve_audio_cache(), tr, bk, fl)
        try:
            st = p.stat()
        except OSError:
            return _err("not found", 404)
        if not p.is_file() or st.st_size == 0:
            return _err("not found", 404)
        return _stream_range(request, p, st.st_size)

    return router


# HTTP Range support. The browser <audio> tag fires a Range request whenever
# the user seeks, so without this every seek triggers a full re-download.
_RANGE_RE = re.compile(r"^bytes=(\d*)-(\d*)$")


def _stream_range(request: Request, file_path: Path, size: int) -> Response:
    range_header = request.headers.get("range")
    if not range_header:
        return _full_response(file_path, size)

    match = _RANGE_RE.match(range_header.strip())
    if not match:
        return Response(
            status_code=416,
            headers={"content-range": f"bytes */{size}", "accept-ranges": "bytes"},
        )
    start_raw, end_raw = match.group(1), match.group(2)
    if start_raw == "" and end_raw != "":
        # suffix-byte-range: last N bytes
        try:
            suffix = int(end_raw)
        except ValueError:
            return _range_not_satisfiable(size)
        if suffix <= 0:
            return _range_not_satisfiable(size)
        start = max(0, size - suffix)
        end = size - 1
    else:
        try:
            start = 0 if start_raw == "" else int(start_raw)
            end = size - 1 if end_raw == "" else int(end_raw)
        except ValueError:
            return _range_not_satisfiable(size)

    if start > end or start < 0 or end >= size:
        return _range_not_satisfiable(size)

    length = end - start + 1
    return StreamingResponse(
        _file_iter(file_path, start, length),
        status_code=206,
        media_type="audio/mpeg",
        headers={
            "accept-ranges": "bytes",
            "content-range": f"bytes {start}-{end}/{size}",
            "content-length": str(length),
        },
    )


def _range_not_satisfiable(size: int) -> Response:
    return Response(
        status_code=416,
        headers={"content-range": f"bytes */{size}", "accept-ranges": "bytes"},
    )


def _full_response(file_path: Path, size: int) -> StreamingResponse:
    return StreamingResponse(
        _file_iter(file_path, 0, size),
        status_code=200,
        media_type="audio/mpeg",
        headers={"accept-ranges": "bytes", "content-length": str(size)},
    )


def _file_iter(file_path: Path, start: int, length: int):
    def gen():
        remaining = length
        with file_path.open("rb") as f:
            f.seek(start)
            while remaining > 0:
                chunk = f.read(min(CHUNK_SIZE, remaining))
                if not chunk:
                    break
                remaining -= len(chunk)
                yield chunk
    return gen()
